### ISSUE TITLE
kubelet repeatedly crashing with `runtime/cgo: pthread_create failed: Resource temporarily unavailable`

### ISSUE THREAD

**Issue by jimmycuadra** [kubernetes/kubernetes#44518]
**What keywords did you search in Kubernetes issues before filing this one?**: pthread_create; resource temporarily unavailable

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): bug report

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T23:37:53Z", GoVersion:"go1.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1", GitCommit:"b0b7a323cc5a4a2019b2e9520c21c7830b7f708e", GitTreeState:"clean", BuildDate:"2017-04-03T20:33:27Z", GoVersion:"go1.8rc2", Compiler:"gc", Platform:"linux/arm"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Raspberry Pi 3
- **OS**: Raspbian GNU/Linux 8 (jessie)
- **Kernel**: Linux 4.4.50-v7+ armv7l GNU/Linux
- **Install tools**: kubeadm 1.6.1

**What happened**:

I have [Pi-hole](https://pi-hole.net/) running on one of my nodes. Here are the resources it consists of:

``` yaml
---
apiVersion: "v1"
kind: "Secret"
metadata:
  name: "pi-hole"
data:
  web-password: "REDACTED" 
type: "Opaque"
---
apiVersion: "v1"
kind: "Service"
metadata:
  name: "pi-hole"
spec:
  selector:
    app: "pi-hole"
    component: "pi-hole"
  ports:
    - port: 80
---
apiVersion: "apps/v1beta1"
kind: "Deployment"
metadata:
  name: "pi-hole"
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: "pi-hole"
        component: "pi-hole"
    spec:
      nodeSelector:
        pi-hole: "true"
      hostNetwork: true
      containers:
        - name: "pi-hole"
          image: "diginc/pi-hole:arm"
          env:
            - name: "ServerIP"
              value: "10.0.1.103"
            - name: "WEBPASSWORD"
              valueFrom:
                secretKeyRef:
                  name: "pi-hole"
                  key: "web-password"
            - name: "DNS1"
              value: "8.8.8.8"
            - name: "DNS2"
              value: "208.67.222.222"
            - name: "VIRTUAL_HOST"
              value: "pi-hole"
          ports:
            - name: "dns-tcp"
              containerPort: 53
              hostPort: 53
              protocol: "TCP"
            - name: "dns-udp"
              containerPort: 53
              hostPort: 53
              protocol: "UDP"
            - name: "http"
              containerPort: 80
              hostPort: 80
          securityContext:
            capabilities:
              add:
                - "NET_ADMIN"
          volumeMounts:
            - name: "etc-pihole"
              mountPath: "/etc/pihole"
      volumes:
        - name: "etc-pihole"
          hostPath:
            path: "/etc/pihole"
```

Every few minutes, the kubelet on this node crashes and the node switches to NotReady status. Here is an excerpt of the kubelet's logs from journald where you can see the initial crash and then another shortly afterwards:

https://gist.github.com/jimmycuadra/0976a12036f709201ae3f937bdd0bb7f

**What you expected to happen**:

Kubelet does not crash.

**How to reproduce it** (as minimally and precisely as possible):

* Create a Kubernetes cluster on Raspberry Pi 3s with the latest Raspbian Jessie Lite.
* `kubectly apply` the YAML file shown above.
* Wait.

**Anything else we need to know**:

This also happened to me on a cluster using HypriotOS 1.4.0. I thought it might have been an issue with Hypriot, so I rebuilt my cluster on regular Raspbian Jessie Lite, but I'm still seeing the same problem. I've also seen it happening on another node where I run a deployment of Traefik.

**Comment by kargakis** [comment]
@kubernetes/sig-node-bugs

**Comment by jimmycuadra** [comment]
Anyone have any ideas about this? @luxas maybe?

**Comment by darktall** [comment]
I am facing same issue.

kubernetes version 1.5.6

docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 49
Server Version: 1.12.6
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp selinux
Kernel Version: 4.11.11-coreos
Operating System: Container Linux by CoreOS 1409.7.0 (Ladybug)

kubelet crash log

Sep 13 10:22:23 coreos systemd[1]: Starting Kubernetes Kubelet...
Sep 13 10:22:23 coreos curl[3831]:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
Sep 13 10:22:23 coreos curl[3831]:                                  Dload  Upload   Total   Spent    Left  Speed
Sep 13 10:22:24 coreos curl[3831]: [158B blob data]
Sep 13 10:22:24 coreos systemd[1]: Started Kubernetes Kubelet.
Sep 13 10:22:24 coreos kubelet[929]: runtime/cgo: pthread_create failed: Resource temporarily unavailable
Sep 13 10:22:24 coreos kubelet[929]: SIGABRT: abort
Sep 13 10:22:24 coreos kubelet[929]: PC=0x7f9e08d38f47 m=5
Sep 13 10:22:24 coreos kubelet[929]: goroutine 0 [idle]:
Sep 13 10:22:24 coreos kubelet[929]: goroutine 34 [running]:
Sep 13 10:22:24 coreos kubelet[929]: runtime.systemstack_switch()
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/asm_amd64.s:252 fp=0xc42045e528 sp=0xc42045e520
Sep 13 10:22:24 coreos kubelet[929]: runtime.gcMarkTermination()
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/mgc.go:1264 +0x368 fp=0xc42045e720 sp=0xc42045e528
Sep 13 10:22:24 coreos kubelet[929]: runtime.gcMarkDone()
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/mgc.go:1150 +0x21e fp=0xc42045e748 sp=0xc42045e720
Sep 13 10:22:24 coreos kubelet[929]: runtime.gcBgMarkWorker(0xc420016a00)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/mgc.go:1497 +0x2c2 fp=0xc42045e7b8 sp=0xc42045e748
Sep 13 10:22:24 coreos kubelet[929]: runtime.goexit()
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc42045e7c0 sp=0xc42045e7b8
Sep 13 10:22:24 coreos kubelet[929]: created by runtime.gcBgMarkStartWorkers
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/runtime/mgc.go:1341 +0x98
Sep 13 10:22:24 coreos kubelet[929]: goroutine 1 [runnable, locked to thread]:
Sep 13 10:22:24 coreos kubelet[929]: k8s.io/kubernetes/vendor/github.com/go-openapi/swag.(*NameProvider).GetJSONNames(0xc420021e30, 0x2a7
Sep 13 10:22:24 coreos kubelet[929]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/go-o
Sep 13 10:22:24 coreos kubelet[929]: k8s.io/kubernetes/vendor/github.com/go-openapi/spec.(*Schema).UnmarshalJSON(0xc42048fb00, 0xc42019b9
Sep 13 10:22:24 coreos kubelet[929]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/go-o
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).object(0xc42000dd10, 0x290b420, 0xc42048fb00, 0x199)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:562 +0x1cb5
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).value(0xc42000dd10, 0x290b420, 0xc42048fb00, 0x199)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:368 +0x2cc
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).object(0xc42000dd10, 0x25eb3a0, 0xc42048fa60, 0x195)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:696 +0x1080
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).value(0xc42000dd10, 0x25eb3a0, 0xc42048fa60, 0x195)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:368 +0x2cc
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).object(0xc42000dd10, 0x23bc380, 0xc42048f8c8, 0x16)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:696 +0x1080
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).value(0xc42000dd10, 0x23bc380, 0xc42048f8c8, 0x16)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:368 +0x2cc
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.(*decodeState).unmarshal(0xc42000dd10, 0x23bc380, 0xc42048f8c8, 0x0, 0x0)
Sep 13 10:22:24 coreos kubelet[929]:         /usr/local/go/src/encoding/json/decode.go:169 +0x1fa
Sep 13 10:22:24 coreos kubelet[929]: encoding/json.Unmarshal(0xc42019b890, 0x235, 0x8570, 0x23bc380, 0xc42048f8c8, 0x40, 0x0)

**Comment by yujuhong** [comment]
Check the task limit for the kubelet service. They may not be sufficient.
```
$ sudo systemctl status kubelet.service
● kubelet.service - Kubernetes kubelet
   Loaded: loaded (/etc/systemd/system/kubelet.service; disabled; vendor preset: disabled)
   Active: active (running) since Wed 2017-09-13 00:06:17 UTC; 17h ago
 Main PID: 1542 (kubelet)
    Tasks: 19 (limit: 4915)
```

**Comment by jimmycuadra** [comment]
FWIW, I've only seen this happening when a pod is using host networking/exposed as a port on the host. Could be a red herring.

**Comment by darktall** [comment]
I am running like 5 pods in a worker node

**Comment by darktall** [comment]
I have three different kubernetes versions running  1.2, 1.5.2, 1.5.6 with same coreos version. i am not seeing this issue on 1.2 ,1.5.2. But only seeing on 1.5.6 version. kubelet  start up configuration is same for all three clusters.

**Comment by luxas** [comment]
@jimmycuadra Can this be reproduced with v1.8?

**Comment by jimmycuadra** [comment]
Trying it again. The pod has been running for an hour so far without the problem coming up. This is on k8s 1.8.1. I'll report back if I see it recur in the next few hours/days.

**Comment by jimmycuadra** [comment]
The pod's been running for 4 days now and I haven't seen any downtime. Whatever it was seems to have been resolved, at least in my case.

**Comment by luxas** [comment]
Great, thanks!

**Comment by shenshouer2916** [comment]
I met the same problem, and I found kubelet process limits like:
![image](https://github.com/user-attachments/assets/1168404f-747a-4fbf-87b0-397be6630d73)
`Max locked memory` was 65536 bytes.

I changed `/etc/security/limits.conf` and add follow content:
```
* hard memlock unlimited
* soft memlock unlimited
```
1. then restart kubelet and found `Max locked memory` not changed
2. `ulimit -l unlimited` and then restart kubelet, `Max locked memory` not changed
3. I changed `/etc/systemd/system.conf` and `/etc/systemd/system.conf` add `DefaultLimitMEMLOCK=infinity` then restart kubelet, `Max locked memory` not changed
4. I changed `/etc/systemd/system/kubelet.service` 
```
[Service]
LimitMEMLOCK=infinity
...
```
the restart kubelet, `Max locked memory` for kubelet was changed.
